__[graphdb and networkx](https://github.com/apowers313/roc/blob/master/experiments/2024.08.09-07.20.42-EXP-graphdb-and-networkx/2024.08.09-07.20.42-EXP-graphdb-and-networkx.ipynb)__

_Experiment Template v0.3.0_

# System Information

## Basic Info

In [1]:
%cd /home/apowers/Projects/roc
!date

/home/apowers/Projects/roc
Thu Jan 23 08:02:27 PST 2025


## GIT

In [2]:
!echo "git hash:" `git rev-parse HEAD`
!echo "git status:"
!git status -s | cat
!echo

git hash: a8b2a56580f878825098f1147b739a15a464cabd
git status:
 M experiments/2024.08.09-07.20.42-EXP-graphdb-and-networkx/2024.08.09-07.20.42-EXP-graphdb-and-networkx.ipynb
 M experiments/2025.01.06-16.50.22-pydantic-fields/2025.01.06-16.50.23-pydantic-fields.ipynb
 M roc/jupyter/save.py
 M tests/graphdb_test.py
?? graph-2025.01.23-07.42.40.gexf
?? graph-2025.01.23-07.43.01.gexf
?? graph-2025.01.23-07.43.23.gml
?? roc/transformer.py
?? test1.gexf
?? test1.gml
?? threadtest.py



## Python

In [3]:
!python --version
!pip --version
!echo "installed packages:"
!pip list

Python 3.13.1
pip 24.2 from /home/apowers/.local/lib/python3.10/site-packages/pip (python 3.10)
installed packages:
Package                            Version          Editable project location
---------------------------------- ---------------- -----------------------------
absl-py                            1.4.0
ale-py                             0.8.1
annotated-types                    0.7.0
anyio                              4.8.0
argon2-cffi                        23.1.0
argon2-cffi-bindings               21.2.0
arrow                              1.2.3
asttokens                          2.2.1
async-lru                          2.0.4
attrs                              23.1.0
Automat                            20.2.0
AutoROM                            0.6.1
AutoROM.accept-rom-license         0.6.1
Babel                              2.12.1
backcall                           0.2.0
bcrypt                             3.2.0
beautifulsoup4                     4.12.2
black                

## System

In [4]:
!uname -a
!echo
!echo "system cpus:"
!lscpu
!echo
!echo "system memory:"
!lsmem
!free

Linux dev.ato.ms 5.15.0-125-generic #135-Ubuntu SMP Fri Sep 27 13:53:58 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

system cpus:
Architecture:             x86_64
  CPU op-mode(s):         32-bit, 64-bit
  Address sizes:          39 bits physical, 48 bits virtual
  Byte Order:             Little Endian
CPU(s):                   32
  On-line CPU(s) list:    0-31
Vendor ID:                GenuineIntel
  Model name:             Intel(R) Core(TM) i9-14900KF
    CPU family:           6
    Model:                183
    Thread(s) per core:   2
    Core(s) per socket:   24
    Socket(s):            1
    Stepping:             1
    CPU max MHz:          6000.0000
    CPU min MHz:          800.0000
    BogoMIPS:             6374.40
    Flags:                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge m
                          ca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 s
                          s ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc 
                          a

# ROC Setup

## ROC Config & Init

In [5]:
import roc

!echo ".env file:"
!cat .env
!echo

roc.init()
# or:
# roc.init({"setting": val})

print(f"config:\n{roc.Config.get()}")  # noqa: T201

.env file:
log_modules="gymnasium:INFO;graphdb:TRACE"
# must have prefix to parse var: https://github.com/pydantic/pydantic-settings/issues/510
roc_expmods=["actions"]


2025-01-23 08:02:31.382 | INFO     | roc.gymnasium:__init__:35 - Gym options: {'options': ['autopickup', 'color', 'disclose:+i +a +v +g +c +o', 'mention_walls', 'nobones', 'nocmdassist', 'nolegacy', 'nosparkle', 'pickup_burden:unencumbered', 'pickup_types:$?!/', 'runmode:teleport', 'showexp', 'showscore', 'time', 'autoopen'], 'max_episode_steps': 100000}


config:
db_host = 127.0.0.1
db_port = 7687
db_conn_encrypted = False
db_username = 
db_password = 
db_lazy = False
db_strict_schema = True
db_strict_schema_warns = False
node_cache_size = 1073741824
edge_cache_size = 1073741824
log_enable = True
log_level = INFO
log_modules = gymnasium:INFO;graphdb:TRACE
gym_actions = (<CompassDirection.N: 107>, <CompassDirection.E: 108>, <CompassDirection.S: 106>, <CompassDirection.W: 104>, <CompassDirection.NE: 117>, <CompassDirection.SE: 110>, <CompassDirection.SW: 98>, <CompassDirection.NW: 121>, <CompassDirectionLonger.N: 75>, <CompassDirectionLonger.E: 76>, <CompassDirectionLonger.S: 74>, <CompassDirectionLonger.W: 72>, <CompassDirectionLonger.NE: 85>, <CompassDirectionLonger.SE: 78>, <CompassDirectionLonger.SW: 66>, <CompassDirectionLonger.NW: 89>, <MiscDirection.UP: 60>, <MiscDirection.DOWN: 62>, <MiscDirection.WAIT: 46>, <MiscAction.MORE: 13>, <Command.ADJUST: 225>, <Command.APPLY: 97>, <Command.ATTRIBUTES: 24>, <Command.CALL: 67>, <Command.CA

## Import Paths

In [6]:
import sys

sys.path.append("./tests")

# Experiment

## NetworkX Export

In [7]:
from roc.graphdb import GraphDB

G = GraphDB.to_networkx()

2025-01-23 08:02:31.419 | TRACE    | roc.graphdb:raw_fetch:116 - raw_fetch: 'MATCH (n) RETURN id(n) as id' *** with params: *** '{}
2025-01-23 08:02:31.424 | TRACE    | roc.graphdb:raw_fetch:116 - raw_fetch: '
                MATCH (src)-[e*0..1]-() 
                WITH src, head(e) AS e
                WHERE id(src) IN [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127]
                RETURN src AS n, collect(e) AS edges
                ' *** with params: *** '{}
2025-01-23 08:02:31.510 | TRACE    | roc.graphdb:find:1003 - find result: {'n': <mgclient.Node(id=125, labels={'Location'}, properties={'name': 'Storms End'}) at 0x7ff1ac0817a0>

In [8]:
G.nodes(data=True)

NodeDataView({125: {'name': 'Storms End', 'labels': 'Location'}, 53: {'name': 'Vardis Egen', 'labels': 'Character'}, 17: {'name': 'House Arryn', 'labels': 'Allegiance'}, 484: {'order': 32, 'labels': 'Death'}, 15: {'name': 'Kings Landing', 'labels': 'Location'}, 43: {'name': 'Jory Cassel', 'labels': 'Character'}, 11: {'name': 'House Stark', 'labels': 'Allegiance'}, 42: {'name': 'Lannister soldier', 'labels': 'Character'}, 476: {'order': 24, 'labels': 'Death'}, 478: {'order': 26, 'labels': 'Death'}, 479: {'order': 27, 'labels': 'Death'}, 480: {'order': 28, 'labels': 'Death'}, 41: {'name': 'Rodrik Cassel', 'labels': 'Character'}, 38: {'name': 'Tribesman', 'labels': 'Character'}, 471: {'order': 19, 'labels': 'Death'}, 472: {'order': 20, 'labels': 'Death'}, 534: {'order': 82, 'labels': 'Death'}, 36: {'name': 'Stark soldier', 'labels': 'Character'}, 465: {'order': 13, 'labels': 'Death'}, 469: {'order': 17, 'labels': 'Death'}, 474: {'order': 22, 'labels': 'Death'}, 475: {'order': 23, 'labels'

In [9]:
import networkx as nx

nx.write_gml(G, "test1.gml")

In [10]:
nx.write_gexf(G, "test1.gexf")

## %save

In [11]:
import roc

roc.init()

/home/apowers/Projects/roc/roc/config.py:139: ConfigInitWarning: Config already initialized, returning existing configuration.
  warnings.warn(
2025-01-23 07:41:53.336 | INFO     | roc.gymnasium:__init__:35 - Gym options: {'options': ['autopickup', 'color', 'disclose:+i +a +v +g +c +o', 'mention_walls', 'nobones', 'nocmdassist', 'nolegacy', 'nosparkle', 'pickup_burden:unencumbered', 'pickup_types:$?!/', 'runmode:teleport', 'showexp', 'showscore', 'time', 'autoopen'], 'max_episode_steps': 100000}


Exception: ExpMod.register attempting to register duplicate name 'random' for module 'action'

In [11]:
%save

2025-01-23 08:03:16.638 | TRACE    | roc.graphdb:raw_fetch:116 - raw_fetch: 'MATCH (n) RETURN id(n) as id' *** with params: *** '{}


Saving 2851 nodes...


Nodes: 100%|███████████████████████████| 2851/2851 [00:00<00:00, 54700.97node/s]


Writing graph to 'graph-2025.01.23-08.03.16'...
!!! debug
writing gexf
Saved 2851 nodes and 12054 edges. Elapsed time: 0:00:00.196847


In [5]:
%save -f gml --no-timestamp foo

format gml
timestamp False
filename foo


In [3]:
%save

Saving 2677 nodes...


Nodes: 100%|███████████████████████████| 2677/2677 [00:00<00:00, 11594.90node/s]

Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.235154


## Performance

### Before Node.get_many()
Saving 2677 nodes...
100%|██████████| 2677/2677 [01:49<00:00, 24.34it/s]
Saved 2677 nodes. Elapsed time: 0:01:49.975708

### After Node.get_many()
Saving 2677 nodes...
Nodes: 100%|██████████| 2677/2677 [00:03<00:00, 756.41node/s] 
Saved 2677 nodes. Elapsed time: 0:00:03.546128

## Formats

"gexf", "gml", "dot", "graphml", "json", "leda", "pajek", "matrix-market", "adj-list", "multi-adj-list", "edge-list"

In [1]:
import roc

roc.init()

2025-01-23 07:44:47.812 | INFO     | roc.gymnasium:__init__:35 - Gym options: {'options': ['autopickup', 'color', 'disclose:+i +a +v +g +c +o', 'mention_walls', 'nobones', 'nocmdassist', 'nolegacy', 'nosparkle', 'pickup_burden:unencumbered', 'pickup_types:$?!/', 'runmode:teleport', 'showexp', 'showscore', 'time', 'autoopen'], 'max_episode_steps': 100000}


FileNotFoundError: could not load experiment modules: ['actions']

In [14]:
%save -f gexf

2025-01-23 07:43:00.963 | TRACE    | roc.graphdb:raw_fetch:116 - raw_fetch: 'MATCH (n) RETURN id(n) as id' *** with params: *** '{}


Saving 2851 nodes...


Nodes: 100%|███████████████████████████| 2851/2851 [00:00<00:00, 56065.12node/s]

Writing graph to 'graph-2025.01.23-07.43.01'...
Saved 2851 nodes and 12054 edges. Elapsed time: 0:00:00.144096


In [12]:
%save -f gml

2025-01-23 08:03:37.028 | TRACE    | roc.graphdb:raw_fetch:116 - raw_fetch: 'MATCH (n) RETURN id(n) as id' *** with params: *** '{}


Saving 2851 nodes...


Nodes: 100%|███████████████████████████| 2851/2851 [00:00<00:00, 49558.66node/s]

Writing graph to 'graph-2025.01.23-08.03.37'...
!!! debug
Saved 2851 nodes and 12054 edges. Elapsed time: 0:00:00.103422


In [4]:
%save -f dot

Saving 2677 nodes...


Nodes: 100%|████████████████████████████| 2677/2677 [00:00<00:00, 9475.87node/s]


Writing graph to 'graph-2024.08.11-16.19.16'...
Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.838471


In [5]:
%save -f graphml

Saving 2677 nodes...


Nodes: 100%|███████████████████████████| 2677/2677 [00:00<00:00, 15400.54node/s]

Writing graph to 'graph-2024.08.11-16.19.16'...


Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.658591


In [6]:
%save -f json-node-link

Saving 2677 nodes...


Nodes: 100%|███████████████████████████| 2677/2677 [00:00<00:00, 15227.67node/s]

Writing graph to 'graph-2024.08.11-16.19.17'...


Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.349104


In [7]:
%save -f json-adj

Saving 2677 nodes...


Nodes: 100%|███████████████████████████| 2677/2677 [00:00<00:00, 15630.61node/s]

Writing graph to 'graph-2024.08.11-16.19.17'...


Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.327570


In [8]:
%save -f cytoscape

Saving 2677 nodes...


Nodes: 100%|███████████████████████████| 2677/2677 [00:00<00:00, 15616.91node/s]

Writing graph to 'graph-2024.08.11-16.19.18'...


Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.465477


In [9]:
%save -f pajek

Saving 2677 nodes...


Nodes: 100%|███████████████████████████| 2677/2677 [00:00<00:00, 15700.66node/s]

Writing graph to 'graph-2024.08.11-16.19.18'...



/home/apowers/Projects/roc/.venv/lib/python3.11/site-packages/networkx/readwrite/pajek.py:75: UserWarning: Node attribute order is not processed. Non-string attribute.
  warnings.warn(
/home/apowers/Projects/roc/.venv/lib/python3.11/site-packages/networkx/readwrite/pajek.py:75: UserWarning: Node attribute imdb_rating is not processed. Non-string attribute.
  warnings.warn(
/home/apowers/Projects/roc/.venv/lib/python3.11/site-packages/networkx/readwrite/pajek.py:75: UserWarning: Node attribute number is not processed. Non-string attribute.
  warnings.warn(
/home/apowers/Projects/roc/.venv/lib/python3.11/site-packages/networkx/readwrite/pajek.py:93: UserWarning: Edge attribute count is not processed. Non-string attribute.
  warnings.warn(


Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.265294


In [10]:
%save -f matrix-market

Saving 2677 nodes...


Nodes: 100%|███████████████████████████| 2677/2677 [00:00<00:00, 15293.06node/s]

Writing graph to 'graph-2024.08.11-16.19.19'...


Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.384356


In [11]:
%save -f adj-list

Saving 2677 nodes...


Nodes: 100%|████████████████████████████| 2677/2677 [00:00<00:00, 9566.32node/s]

Writing graph to 'graph-2024.08.11-16.19.19'...
Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.294720


In [12]:
%save -f multi-adj-list

Saving 2677 nodes...


Nodes: 100%|███████████████████████████| 2677/2677 [00:00<00:00, 15098.70node/s]

Writing graph to 'graph-2024.08.11-16.19.19'...


Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.207788


In [13]:
%save -f edge-list

Saving 2677 nodes...


Nodes: 100%|███████████████████████████| 2677/2677 [00:00<00:00, 14888.47node/s]

Writing graph to 'graph-2024.08.11-16.19.20'...


Saved 2677 nodes and 11967 edges. Elapsed time: 0:00:00.223435
